### ISSUE TITLE
Nested multiple connections problems.

### ISSUE THREAD

**Issue by ktanakaj** [fntneves/laravel-transactional-events#23]
Hi! I have many databases in one app. But there's something wrong with laravel-transactional-events when nested multiple connections. I report these cases.

### OK (Single connection)
```
\DB::connection('logdb')->transaction(function() {
    event(new \App\Events\EventA);
    event(new \App\Events\EventB);
});
```
Of course it is no problems. I got both events.

### NG (Exception when nested multiple connections)
```
\DB::connection('logdb')->transaction(function() {
    \DB::transaction(function() {
        event(new \App\Events\EventA);
    });
    event(new \App\Events\EventB);
});
```
I got exception below.
```
ErrorException: count(): Parameter must be an array or an object that implements Countable in /vagrant/server/vendor/fntneves/laravel-transactional-events/src/Neves/Events/TransactionalDispatcher.php:152
Stack trace:
#0 [internal function]: Illuminate\Foundation\Bootstrap\HandleExceptions->handleError(2, 'count(): Parame...', '/vagrant/server...', 152, Array)
#1 /vagrant/server/vendor/fntneves/laravel-transactional-events/src/Neves/Events/TransactionalDispatcher.php(152): count(NULL)
```

### NG (Inner event not fired when nested multiple connections)
```
\DB::connection('logdb')->transaction(function() {
    event(new \App\Events\EventB);
    \DB::transaction(function() {
        event(new \App\Events\EventA);
    });
});
```
I got EventB only. EventA is not fired.

### My Environment
* laravel 5.8.10
* laravel-transactional-events 1.8.0
* CentOS 7
* MySQL 8.0

**Comment by ktanakaj** [comment]
I think there are some solutions.

1. All events fire after root transaction committed.
2. Each events fire after each transaction that event including committed.
3. Add `$connection` properties to `TransactionalEvent` like `Model`, and each events work for own connection.

**Comment by ktanakaj** [comment]
I might fixed my local source in the following way.

* Add `$connection` properties to `TransactionalEvent`.
* `dispatch` method use `$connection` if the event has it.
* `$this->transactionLevel` change to `$this->transactionLevels[$connection->getName()]`.

I don't know whether this is the best...

**Comment by fntneves** [comment]
Hi @ktanakaj,

Will take a look into this issue. Thank you for opening it!

**Comment by azeos** [comment]
I'm getting `"count(): Parameter must be an array or an object that implements Countable"` too.

### Environment:
* Laravel 5.7.21
* PHP 7.2
* laravel-transactional-events 1.4.6

I think the problem has something to do with [PHP 7.2](https://php.net/manual/en/migration72.incompatible.php#migration72.incompatible.warn-on-non-countable-types)

[TransactionalDispatcher.php](https://github.com/fntneves/laravel-transactional-events/blob/1.4.6/src/Neves/Events/TransactionalDispatcher.php#L152)
```PHP
protected function addPendingEvent($connection, $event, $payload)
{
    $connectionId = $connection->getName();
    $transactionLevel = $this->transactionLevel;
    $eventData = [
        'event' => $event,
        'payload' => is_object($payload) ? clone $payload : $payload,
    ];
    $transactionLevelEvents = &$this->pendingEvents[$connectionId][$transactionLevel];
    $transactionLevelEvents[count($transactionLevelEvents) - 1][] = $eventData;
}
```
In my case `$transactionLevelEvents` is `null`, that's triggering the exception.

**Comment by fntneves** [comment]
@azeos Can you please show an example snippet that raises this exception?

Is it the same as @ktanakaj showed above?

**Comment by azeos** [comment]
@fntneves I updated my comment. Could `$transactionLevelEvents` be null? Or something is wrong before that?

**Comment by azeos** [comment]
@fntneves any updates on this?

**Comment by fntneves** [comment]
@azeos I am working on this issue. Hopefully, will push a new release during this weekend.

**Comment by azeos** [comment]
@fntneves thanks! Just to know if I need to search for another alternative.

**Comment by fntneves** [comment]
@azeos Just to contextualise, the solution of this issue is more complex than it seems to be.

I must consider _SAVEPOINTS_ scenarios, that is, nested transactions within the same connection and also force that events are dispatched the same order they were registered. 

Give me some time to organise this, most of the code will be changed.

**Comment by fntneves** [comment]
@azeos I will follow one of the proposed solutions from @ktanakaj. Events will be triggered only after the root transaction commits, which means that if the inner transaction succeeds, if the root transaction fails, no event is fired at all.  For such reason, I do not recommend to rely on events dispatched on nested transactions on distinct database services.

In detail, the concept of Transactional Event is not suitable for transactions of distinct connections, because, in the end, they are not effectively related. Considering the example given by @ktanakaj, if the transaction on **logdb** connection fails to commit, it will not rollback the inner transaction. At this point, should we fire the events fired by the inner transaction? If yes, application could end in an inconsistent state, because these events should not be dispatched before the previous ones (see example below, where _LogConnectionEvent()_ is registered after _CustomEvent()_). Therefore, it makes no sense to handle transactions of distinct connections as if they were related, except if anything like distributed transactions (XA) is adopted.

```php
DB::transaction(function(){
    event(new CustomEvent());

    DB::connection('log')->transaction(function() {
        event(new LogConnectionEvent());
    });

    event(new SecondCustomEvent());
})
```

I will update this package to fix the bug on using custom connections, but it will only dispatch events after the root transaction commits, even if the inner transaction commits.

**Comment by azeos** [comment]
@fntneves thanks for the update on this subject. I understand what you are saying and it seems reasonable.

**Comment by fntneves** [comment]
Could you please check version 1.8.1?

**Comment by ktanakaj** [comment]
Thanks a lot! I confirmed version 1.8.1.
The events were fired after root transaction committed.
Looks Good To Me :thumbsup:

**Comment by fntneves** [comment]
@azeos Any feedback on this?

**Comment by azeos** [comment]
@fntneves I'll try tomorrow, I'm working on another project right now. Will ley you know as soon as I try it. Thanks a lot!

**Comment by azeos** [comment]
@fntneves today I went back with this problem. Unfortunately I'm using Laravel 5.7, so I can't test v1.8.1, sorry, didn't pay attention. An update to 5.8 is planned, so I'll be back.

Thanks!

**Comment by fntneves** [comment]
@azeos I'm applying the fix for 1.4.x, also.

Will release a new 1.4.x version today. :+1:

**Comment by azeos** [comment]
@fntneves just to let you know that I updated to Laravel 6 and it's working great. Thanks for the package!
